### DISCUSSION TITLE
(Re)naming pattern composers

### DISCUSSION THREAD

**Discussion by yaxu** [tidalcycles/strudel#82] [Ideas]
There are already a few ways to compose two patterns together by matching events and merging their values, `union`, `add`, `mul`, `sub`, and `add`. 

`union` isn't a good name for this, because they all do a union. The distinguishing feature is that when doing the union, `a.union(b)` it replaces values from pattern `a` with values from `b`. So I propose renaming it to `set` (as in, set the values to, not set theory).

A variant that keeps values in `a` could be called `const`.

Where matching events don't have the same timespan but do overlap, currently you end up with the intersection as a fragment of the pattern on the left (pattern `a` in the above). We can also have variants `setRight`, `addRight`, `mulRight` etc. This is a bit ambiguous though, especially for`constRight`, because there's confusion about whether it's the fragment structure, or the value that we're taking from the rightmost pattern. Maybe `setFlip`, `constFlip` etc for 'flipping' the structure is better? This assumes that taking wholes from the left is default.

Then there is the variant where the intersections of matching events are treated as whole events. E.g. where `"1 2 3"` + `"10 20"` gives `11 [12 22] 23`. One possibility is using the suffix 'Sect' for intersection. `setSect`, `addSect` etc. I quite like this because you can also think about it dividing events into sections.

(In Tidal, this is done with operators like `|>` for `set`, `>|` for `setFlip` `|<` for `const`, `|<|` for `constSect`, `|+` for `add`, `+|` for `addFlip`, `+` or `|+|` for `addSect`, etc. This works nicely, but because javascript doesn't support custom operators, and there aren't enough to overload, we have to find a reasonable way to do this with words instead. Still, it would be nice to alias `+` for `add`)

All of the variants so far assume that we're composing patterns by aligning cycles and seeing which events intersect, then combining them. There are more possibilities though! For example, squeeze, which aligns cycles on the right with event timespans on the left.

```js
"1 2 3".addSqueeze("10 20")
===
"[11 21] [12 22] [13 23]"
```
... and then we might want to flip and squeeze:
```js
"1 2 3".addSqueezeFlip("10 20")
===
"[11 12 13] [21 22 23]"
```

I think there could be a lot of uncharted possibilities for squeeze.. and probably other structural ways to combine patterns.

**Comment by yaxu** [comment]
Implemented in https://github.com/tidalcycles/strudel/pull/83

**Comment by yaxu** [comment]
Merged

**Comment by felixroos** [comment]
this might be another way to combine patterns https://github.com/tidalcycles/strudel/issues/53

**Comment by felixroos** [comment]
This looks all really good. Let me write down some lists to help my brain a bit: 

**1. there are 2 main aspects of `haps`:**

- a. `value`: `hap` content
- b. `structure`: temporal structure, where the haps land in time

**2. strategies to merge `values`** 

- a. `add` / `mul` / `sub` / `div`: binary arithmetic operations, `"0 1".add("1 2") === "1 3"`
- b. `set`: assign to object `values`: `n("0 1").set(room(.5))` -> `{n:0,room:.5}, {n:1,room:.5}`
- c. `keep`: keep value as is, only merge `structure`
- what else?

There are shorthand pattern methods to `set` for each value type: `n("0 1").room(.5) === n("0 1").set(room(.5))`. Here `.room` is the shorthand.

**3. strategies to merge `structure`** 

- a. '' (empty): use structure from the left
- b. `flip`: use structure from the right
- c. `sect`: merge structure from left and right
- d. `reset`
- e. `restart`

for `reset` and `restart`, see https://github.com/tidalcycles/strudel/issues/53 + [discord discussion](https://discord.com/channels/779427371270275082/937365093082079272/967142987719143444)

**4. pattern composers = combination of 2 and 3**

Using both strategies for `values` and `structure`, we can generate all possible ways of pattern interaction:

`<valueStrategy><structureStrategy>`

| . | add | set | keep |
| - | --- | -- | --- |
| '' | .add | .set | .keep |
| flip | .addFlip | .setFlip | .keepFlip |
| sect | .addSect | .setSect | .keepSect |
| reset | .addReset | .setReset | .keepReset |
| restart | .addRestart | .setRestart | .keepRestart |

**5. Thoughts**

I'm not sure if I like the `''` / `'Flip'` / `'Sect'` naming. For my understanding `''` (or `'Left'`) / `'Right'` / `'Both'` would be more logical. Also, the `sect` is pretty close to `set`.

| . | add | set | keep |
| - | --- | -- | --- |
| '' | .add | .set | .keep |
| left | .addLeft | .setLeft | .keepLeft |
| right | .addRight | .setRight | .keepRight |
| both | .addBoth | .setBoth | .keepBoth |
| reset | .addReset | .setReset | .keepReset |
| restart | .addRestart | .setRestart | .keepRestart |

Also, it would be easier to think about this stuff if we had some better wording for "strategy to merge values" + "stategy to merge structure"

**6. How does `reset` / `restart` / `squeeze` fit in here?**

Are those siblings to '' / `flip` / `sect` ? Or would it make sense to `resetFlip` `resetSect` ?
Edit: okay its a [sibling](https://github.com/tidalcycles/strudel/blob/60027e664f0cae928a743ff11314f18b4122233d/packages/core/pattern.mjs#L807).

**7. default pattern methods for structure?** 

With `reset` and `restart` as part of the `structure` family, the default cases `.reset` and `.restart` don't make much sense

We could use

```js
Pattern.prototype.reset = Pattern.prototype.keepReset;
Pattern.prototype.restart = Pattern.prototype.keepRestart;
```

**Reply by yaxu** [reply]
1. Yep! Two main aspects of haps, and therefore, two main aspects of combining two patterns together.
2. Strategies to merge values:

> a. add / mul / sub / div: binary arithmetic operations, "0 1".add("1 2") === "1 3"

Yes but good to keep in mind that you can't merge values without also merging time. By default, temporal structure comes from the left, so `"0 1".add("3") === "3 4"`

> b. set: assign to object values: n("0 1").set(room(.5)) -> {n:0,room:.5}, {n:1,room:.5}

No `set` is just another way to merge values, it's just `(a,b) => a` (i.e. `id`).

So `set` and `add` (and all the others) do the same where there aren't matching keys:
```js
n("0 1").set(room(.5)) -> {n:0,room:.5}, {n:1,room:.5}
n("0 1").add(room(.5)) -> {n:0,room:.5}, {n:1,room:.5}
```
The 'assigning to object values' just come from the union that is done. `set` also works on patterns of primitives:
```js
"0 1".set(5) -> "5 5"
```
> c. keep: keep value as is, only merge structure

Maybe this is better described as 'keep value from the left', discarding the one on the right. It is merging the values with `(a,b) => b`

3. There is also `squeeze` and `squeezeFlip`

5. I've moved away from left/right/both, because it's ambiguous -- it's not clear whether the left/right refers to value or time. I find myself reading `setRight` as setting the value on the right, rather than setting the value on the left and taking structure from the right.
So instead there's an assumption that everything is relative to the left. Then `set` and `keep` either set the value on the left, or keeps it as-is. '' or `flip` will either take the structure from the left (the default), or flip it round to take it from the right. 
Agreed there probably is a nicer word than `sect`.

> Also, it would be easier to think about this stuff if we had some better wording for "strategy to merge values" + "stategy to merge structure"

Agreed! I've also long struggled with the 'structure from left/right' explanation, as structure always comes from both sides. The difference is in how you deal with fragmenting events when they intersect but don't align perfectly - either by treating them as fragments of events from the left, from the right, or as in new, whole events.

6. Yes they are siblings. But! There is a `squeezeFlip`. This is a slightly different kind of flipping though... Not to do with fragments but which cycles go inside which events. So maybe needs renaming.

7.

> With reset and restart as part of the structure family, the default cases .reset and .restart don't make much sense

I've been wondering about this too, but I think they do make sense. Structuring rhythms separately from values is strange but really interesting:

```js
"[x [~ x] <x x*2> [~ x]]".off("<0.25 0.125>", slow(2)).reset("a3 g3*2 [~ a3] e3".slow(4))
```

**Reply by felixroos** [reply]
> I've moved away from left/right/both, because it's ambiguous -- it's not clear whether the left/right refers to value or time

What if we use this syntax:

```js
"0 1".add("1 2 3") // "1 3" (default = left)
"0 1".left.add("1 2 3") // "1 3"
"0 1".right.add("1 2 3") // "1 2 4"
"0 1".merge.add("1 2 3") // "1 [2 3] 4"
```

**Reply by felixroos** [reply]
Now I also wonder why there is no way to sample the value at the middles or tails:

```js
"0 1".leftEnd.add("1 2 3") // "1 4"
"0 1".rightEnd.add("1 2 3") // "1 3 4"
```

not sure if this is practical at some point, but just something that is "structurally missing", as values are always sampled at the onset, so why not at the middle or ends?

**Reply by felixroos** [reply]
and btw thanks for clarifying! `set` is more versatile than I thought

**Reply by yaxu** [reply]
(moving from separate post)

Thinking a bit more on `resetSet` vs `resetKeep`.. Issue #53 originally called for a 'binary pattern reset function'. Currently we're not driving it with a binary pattern but an 'event-and-silence' pattern.

If we really used a binary pattern, i.e. a pattern of booleans, then 'false' could take the value from the opposite side of 'true'. Then `reset(false, true)` would do the same as `resetKeep(true,false)`. So then `reset("0 [0 0]")` would be shorthand for `resetKeep("1 [1 1]")`.

**Reply by felixroos** [reply]
> Also, it would be easier to think about this stuff if we had some better wording for "strategy to merge values" + "stategy to merge structure"

maybe value combinators and structure combinators

**Reply by yaxu** [reply]
> What if we use this syntax:
> ```js
> "0 1".add("1 2 3") // "1 3" (default = left)
> "0 1".left.add("1 2 3") // "1 3"
> "0 1".right.add("1 2 3") // "1 2 4"
> "0 1".merge.add("1 2 3") // "1 [2 3] 4"
> ```

Hmm, I'm not keen on this. `left` and `right` are still ambiguous, and it's not clear to me what kind of thing a `"0 1".left` would be.

I think `merge` isn't the right name as they all merge patterns and values together.

> Now I also wonder why there is no way to sample the value at the middles or tails

True! Similarly a counterpart to `reset` that played the last part of a cycle, rather than the first part.

**Reply by felixroos** [reply]
> (moving from separate post)
> 
> Thinking a bit more on `resetSet` vs `resetKeep`.. Issue #53 originally called for a 'binary pattern reset function'. Currently we're not driving it with a binary pattern but an 'event-and-silence' pattern.
> 
> If we really used a binary pattern, i.e. a pattern of booleans, then 'false' could take the value from the opposite side of 'true'. Then `reset(false, true)` would do the same as `resetKeep(true,false)`. So then `reset("0 [0 0]")` would be shorthand for `resetKeep("1 [1 1]")`.

maybe its better to not use binary patterns, so you are able to use any pattern as reset pattern, like i did earlier with scales. Otherwise, normal patterns would need to be converted to binary, and sometimes you‘ll also want a zero as value that should not end up false

**Reply by yaxu** [reply]
Yes true, and `resetSet` needs the pattern on the right to have the values to set.

In tidal there are `sew` and `stitch` for merging two patterns with a third binary pattern.

**Reply by felixroos** [reply]
i always wondered how .struct fits in there. many times, I wanted to use .struct while also associate a value with each true value. Wouldnt it make sense to overload struct like that:

"0 1".struct("1 2 3", add) so that would be like addFlip.. maybe also name it structWith, or withStruct..

**Reply by felixroos** [reply]
for sect, this could be named mergeStruct or sth

**Reply by yaxu** [reply]
Hm I don't understand, in `"a b c".struct("t f t")`, the `a` and `c` value is already associated with each true value.

Or do you mean associate a function with each true value?

**Reply by felixroos** [reply]
yes i meant to associate a value that is then combined with the target pattern using a function

**Reply by yaxu** [reply]
I see, have you seen the `when` function?

https://github.com/tidalcycles/strudel/blob/47717e872b87f497990d7c1d6c3d54bfb96f2003/packages/core/pattern.mjs#L580-L587

I think you're looking for this but with `appLeft` instead of `appRight`.. So maybe a `whenFlip`..

**Reply by yaxu** [reply]
Ah no you're not looking for a function driven by a binary pattern.. So `addFlip` indeed.

**Reply by yaxu** [reply]
Binary patterns are great though because you can easily invert them.
`"a".struct("1 0 [1 0]*2 1".every(4, invert))`

and use the same binary pattern across different parts of a pattern, sometimes inverted/shifted, to have a weird shared metrical feel

**Reply by yaxu** [reply]
Relatedly there's also `func`:
```js
"0 [2 3]".func(x => (x + 3)/2, [x => 0 - x, x => 7])
===
"1.5 [-2 7]"
```

**Comment by yaxu** [comment]
Let's collect words for combining two patterns:

| action | options |
| - | --- |
| fragment events from left | '' (default), left, admit, fit, cast, shape, adopt, form |
| fragment events from right | flip, right, render, affect, impinge, fix, adapt, deform |
| intersect events | sect, section, merge, meld, interlace, mix, fragment, rend, cut, blend |
| align sam on left to start of event on right | reset, inhabit, dwell |
| align cycle zero on left to start of event on right | restart |
| fit cycles on right to events on left | squeeze, inhabit |
| fit cycles on left to events on right | squeezeFlip, expand, house |

**Comment by felixroos** [comment]
I'll try to rephrase my idea with struct here:

> "0 1".struct("1 2 3", add) so that would be like addFlip.

I meant, instead of (or in addition to) trying to come up with a single function name with the form 
`<valueStrategy><structureStrategy>`, we could make it more explicit by extending the `struct` function family:
```js
"0 1".add("1 2 3") // structure from left
"0 1".structWith("1 2 3", add) // structure from right
"0 1".mergeWith("1 2 3", add) // merge structure
"0 1".resetWith("1 2 3", add) // reset structure
"0 1".restartWith("1 2 3", add) // restart structure
```

For me, this feels pretty consistent with the existing API design, as many functions come with an added `With`, allowing to pass in a function. This comes with the additional benefit of being able to pass a custom function:

```js
"0 1".structWith("1 2 3", (l,r) => l**r) 
```

**Reply by yaxu** [reply]
Right that seems promising. So then `add` and friends need some adjustment so that they also work on primitives, in which case the 'structure from left' aspect of `add` doesn't apply (as there aren't patterns involved).

I think it might be better this way around:
```js
"0 1".structWith(add, "1 2 3") 
```
So that this could work:
```js
"0 1".structWith(add, 1, 2, 3) 
```
.. and I think it reads better that way as well.

**Reply by yaxu** [reply]
But do we also make/keep `"0 1".structAdd` as shorthand for `"0 1".struct(add, "1 2 3")`? I think the latter is clearer, that structuring and adding are separate things going on. But once you've learned that, maybe it's nice to have the shorthand.
(we also haven't had a proper conversation about camelCase)

**Reply by yaxu** [reply]
Or maybe even `struct` and friends should check if the first argument is a function and the later once aren't, so `"0 1".struct(add, "1 2 3")` could work

**Reply by yaxu** [reply]
prototype for `"0 1".struct(add, "1 2 3")`
```js
const _makeOp = curry(function (func, a, b) {
  if (a instanceof Pattern || b instanceof Pattern) {
    return reify(a).fmap(func).appBoth(reify(b));
  }
  return func(a)(b);
});

export const newAdd = _makeOp(a => b => a + b);

Pattern.prototype.newStruct = function (...args) {
  var func = undefined;
  if (args[0] instanceof Function) {
    func = args.shift();
  }
  else {
    func = constant;
  }
  return this.fmap(func).appLeft(sequence(args));
}
```

**Comment by yaxu** [comment]
So for operation `op` (like add, keep, set, ...) and structuring `struct` (like reset, merge, struct, ...), it could be

1. `a.op(struct, ...b)`
2. `a.struct(op, ...b)`
3. `a.opStruct(...b)`
4. `a.structOp(...b)`
5. `a.op.struct(...b)`
6. `a.struct.op(...b)`

7. Or one more idea, if this is too much to type we could look for initialisms, like

| . | . |
| --- | --- |
| '' | keep structure from left |
| S | restructure with right |
| M | merge structure from both sides |
| Sq | squeeze |
| X | reset |
| Xx | restart |

to make `addSq`, `addM`, etc..

At the moment I prefer 3 or 7, but quite like 1 as well

**Reply by yaxu** [reply]
we could still support `.struct((a,b) => a + b, "1 2 3")`
but for compatibility with tidal
`.struct("true false")`
assumes a binary pattern and `keep`
whereas
`.squeeze("a b")`
works with any pattern and assumes `set`
if we standardise `struct` on `set` as well, it applies both values and structure, and becomes a kind of reverse `mask`..

**Reply by yaxu** [reply]
Lets also consider using `in` and `out` rather than `left` and `right` or `` and `struct`

Then we're either performing an operation to values inside their structure, or taking the values out of their structure.

`addIn`, `addOut`, `keepIn`, `keepOut` (!), and also `addSqueezeIn`/`addSqueezeOut`..

Instead of `merge` or `sect` we could then have `mid` or `tween`.

**Comment by felixroos** [comment]
Not sure where to put this, I just took some notes while rereading `pattern.mjs` after the clean up. 
Happened to compare composers and transforms, which is also related to https://github.com/tidalcycles/strudel/issues/264 and https://github.com/tidalcycles/strudel/issues/36 . (Note that through auto formatting my notes, the double quotes went single)

These are the [composers](https://github.com/tidalcycles/strudel/blob/main/packages/core/pattern.mjs#L1389):

```js
const composers = [
  'set',
  'keep',
  'keepif',
  'add',
  'sub',
  'mul',
  'div',
  'mod',
  'pow',
  '_and',
  '_or',
  '_xor',
  '_lshift',
  '_rshift',
  'lt',
  'gt',
  'lte',
  'gte',
  'eq',
  'eqt',
  'ne',
  'net',
  'and',
  'or',
  'func',
];
```

they are used to programatically define Pattern.prototype methods for different pattern aligment strategies:

```js
['In', 'Out', 'Mix', 'Squeeze', 'SqueezeOut', 'Trig', 'Trigzero'];
```

Example:

```js
'0 1 2 1'.addOut('1 2'); // = "1 3"
'0 1 2 1'.addIn('1 2'); // = "1 2 4 3"
```

then there are these [curried transformation](https://github.com/tidalcycles/strudel/blob/main/packages/core/pattern.mjs#L1781) functions (further called transforms):

```js
const transforms = [
  'add',
  'chop',
  'chunk',
  'chunkBack',
  'div',
  'early',
  'echo',
  'every',
  'fast',
  'inv',
  'invert',
  'iter',
  'iterBack',
  'jux',
  'juxBy',
  'late',
  'linger',
  'mask',
  'mul',
  'off',
  'ply',
  'range',
  'rangex',
  'range2',
  'rev',
  'slow',
  'struct',
  'sub',
  'superimpose',
  'set',
  'when',
];
```

these can be used as transformation methods like `every`, `echoWith`, `superimpose` etc:

```js
'0 1 2'.superimpose(fast(2));
```

Let's look at how the two lists overlap:

```js
composers.filter((c) => transforms.includes(c));
['set', 'add', 'sub', 'mul', 'div'];
```

In this list are functions you can use for both:

```js
note('c3 d3 e3').addOut('1 2'); // works
note('c3 d3 e3').superimpose(add('1 2')); // works
note('c3 d3 e3').superimpose(addOut('1 2')); // what about?
```

These are functions you can only use as composer:

```js
// transforms that are not composers
composers.filter((c) => !transforms.includes(c));
[
  'keep',
  'keepif',
  'mod',
  'pow',
  '_and',
  '_or',
  '_xor',
  '_lshift',
  '_rshift',
  'lt',
  'gt',
  'lte',
  'gte',
  'eq',
  'eqt',
  'ne',
  'net',
  'and',
  'or',
  'func',
];
```

example:

```js
note('c3 d3 e3').powOut('1 2'); // composer: works
note('c3 d3 e3').superimpose(pow(2)); // does not work
```

At last, let's look at functions that can only be used as transforms and not as composers:

```js
// composers that are not transforms
transforms.filter((c) => !composers.includes(c));
[
  'chop',
  'chunk',
  'chunkBack',
  'early',
  'echo',
  'every',
  'fast',
  'inv',
  'invert',
  'iter',
  'iterBack',
  'jux',
  'juxBy',
  'late',
  'linger',
  'mask',
  'off',
  'ply',
  'range',
  'rangex',
  'range2',
  'rev',
  'slow',
  'struct',
  'superimpose',
  'when',
];
```

Example

```js
note('c3 d3 e3').superimpose(fast('<1 2>')); // transform: works
note('c3 d3 e3').fastOut('1 2'); // composer: doesnt work
```

Additionally, there is the large list of [control params](https://github.com/tidalcycles/strudel/blob/main/packages/core/controls.mjs#L10) which would also need composed variants e.g. `.speedOut()`

General questions:

- Which functions should only be (composers | transforms)?
- Which functions should be both composers and transforms?
- Does it make sense to have transforms that also do aligment, like `s("bd sd").superimpose(speedOut("1 2 3"))`

I still think it might be a nice thing to have an alternative (possibly coexisting) to the matrix style function generation. Like this:

```js
note('c3 d3 e3').structWith(add, '1 2');
note('c3 d3 e3').squeezeWith(add, '1 2');
```

(`struct` is used as synonym for `out` here)

With those, we could also solve the transforms that also do alignment thing with:

```js
s("bd sd").superimpose(x=>x.structWith(speed, "1 2 3"))
```

**Reply by yaxu** [reply]
Ah nice!

```js
note('c3 d3 e3').addOut('1 2'); // works
note('c3 d3 e3').superimpose(add('1 2')); // works
```
Do you mean they do the same? The latter will add events to the original (unless there's something wrong in the implementation or my understanding..).

* Which functions should only be (composers | transforms)?
* Which functions should be both composers and transforms?

Reading your notes, I think there isn't/shouldn't be much of a difference.

`"4".pow(2)` and friends _do_ work, with default `set` behaviour. The top-level `pow` function is defined, but should be. As discussed on discord (I think) we should probably do this dynamically, exporting the functions from the module differently to make this possible, like controls.mjs.

One difference between composers and transforms is that 'composers' come from infix operators in Haskell, so have exactly two patterned arguments, including `this`. Transforms can have any arity. The other difference is that 'composers' compose together values, with a default or explicit pattern alignment to make that possible.

But really I think these composers are just a certain kind of transform.. Or we could call all of them composers, because the (patternified) 'transforms' compose their patterned arguments together just like 'composers' do.

* Does it make sense to have transforms that also do aligment, like s("bd sd").superimpose(speedOut("1 2 3"))

I think not like this, because the alignment doesn't really belong to the transform, but each of its arguments.

So say we do away with `powOut` and friends in favour of the new `pow.out`.

That means we can hopefully support `"1 2 3".every(3, rev)` as now, but also e.g. `every.squeeze("3 4").set(rev, fast(2))`. I.e, we can set a different alignment for each parameter.

I'm not sure what that last example ebove would actually do, but it makes me think that we could use `func` there to be explicit about the functional application, like `every(3).func(rev, fast(2))` , making space for just putting a value there like `every(3).set(4)`.. maybe we'd even want to support `every(3).rev(4)`? probably that would get confusing..
